pkg/cmd/cli/cmd/create_route.go

```diff
@@ -0,0 +1,396 @@
+	oc, kc, err := f.Clients()
+	if err != nil {
+		return err
+	}
+
+	ns, _, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	route, err := unsecuredRoute(cmd, kc, ns, args)
+	if err != nil {
+		return err
+	}
+
+	route.Spec.Host = kcmdutil.GetFlagString(cmd, "hostname")
+	route.Spec.Path = kcmdutil.GetFlagString(cmd, "path")
+
+	route.Spec.TLS = new(api.TLSConfig)
+	route.Spec.TLS.Termination = api.TLSTerminationEdge
+
+	cert, err := fileutil.LoadData(kcmdutil.GetFlagString(cmd, "cert-file"))
+	if err != nil {
+		return err
+	}
+	route.Spec.TLS.Certificate = string(cert)
+	key, err := fileutil.LoadData(kcmdutil.GetFlagString(cmd, "key-file"))
+	if err != nil {
+		return err
+	}
+	route.Spec.TLS.Key = string(key)
+	caCert, err := fileutil.LoadData(kcmdutil.GetFlagString(cmd, "ca-cert-file"))
+	if err != nil {
+		return err
+	}
+	route.Spec.TLS.CACertificate = string(caCert)
+	destCACert, err := fileutil.LoadData(kcmdutil.GetFlagString(cmd, "dest-ca-cert"))
+	if err != nil {
+		return err
+	}
+	route.Spec.TLS.DestinationCACertificate = string(destCACert)
+
+	route, err = oc.Routes(ns).Create(route)
+	if err != nil {
+		return err
+	}
+	mapper, typer := f.Object()
+	resourceMapper := &resource.Mapper{
+		ObjectTyper:  typer,
+		RESTMapper:   mapper,
+		ClientMapper: resource.ClientMapperFunc(f.Factory.RESTClient),
+	}
+	info, err := resourceMapper.InfoForObject(route)
+	if err != nil {
+		return err
+	}
+	shortOutput := kcmdutil.GetFlagString(cmd, "output") == "name"
+	kcmdutil.PrintSuccess(mapper, shortOutput, out, info.Mapping.Resource, info.Name, "created")
+	return nil
+}
+
+// unsecuredRoute will return a route with enough info so that it can direct traffic to
+// the service provided by --expose. Callers of this helper are responsible for providing
+// tls configuration, path, and the hostname of the route.
+func unsecuredRoute(cmd *cobra.Command, kc *kclient.Client, namespace string, args []string) (*api.Route, error) {
+	serviceName := kcmdutil.GetFlagString(cmd, "expose")
+	// Right now the service is required to exist in order for oc create route
+	// or oc expose to generate a route. In the future we may want to create
+	// routes prior to service creation.
+	svc, err := kc.Services(namespace).Get(serviceName)
+	if err != nil {
+		return nil, err
+	}
+
+	name := svc.Name
+	switch len(args) {
+	case 0:
+	case 1:
+		name = args[0]
+	default:
+		return nil, fmt.Errorf("multiple names provided. Please specify at most one")
+	}
+
+	var (
+		routePort   util.IntOrString
+		supportsTCP bool
+	)
+
+	for _, port := range svc.Spec.Ports {
+		if port.Protocol == kapi.ProtocolTCP {
+			portString := kcmdutil.GetFlagString(cmd, "port")
+			if len(port.Name) > 0 && len(portString) == 0 {
+				// If the service has multiple ports and the user didn't specify --port,
+				// then default the route port to a service port name.
+				routePort = util.NewIntOrStringFromString(port.Name)
+			} else if len(portString) > 0 {
```

Comment:
This logic is a little hard for me to follow.  I would have expected specifying `--port` wouldn't bother inspecting the the service at all.  I think that's what this code is trying to do, but it only performs an action if there's a TCP port.  Which isn't strange in and of itself, but it weird when coupled to the `supportsTCP` check.

I think it should be something like this and keeping the logic out of the loop will help people read it.
```
if the service doesn't exist and the port wasn't specified, fail
if the port was specified, then assign the routePort accordingly
else if multiple target ports are exposed by the service, fail
else if a single target port exists, assign the routePort
if the service exists and no TCP port exists, fail


```